neqo-transport/src/connection/mod.rs

```diff
@@ -1977,6 +1989,33 @@ impl Connection {
         Ok((tokens, ack_eliciting, padded))
     }
 
+    fn maybe_write_datagram(
+        &mut self,
+        builder: &mut PacketBuilder,
+        tokens: &mut Vec<RecoveryToken>,
+    ) -> bool {
+        if let Some(data) = &self.datagram {
+            let len = data.len();
+            if builder.remaining() >= len + 1 {
+                // + 1 for Frame type
+                let length_len = Encoder::varint_len(u64::try_from(len).unwrap());
+                if builder.remaining() > 1 + length_len + len {
+                    builder.encode_varint(FRAME_TYPE_DATAGRAM_WITH_LEN);
+                    builder.encode_vvec(&data);
+                } else {
+                    builder.encode_varint(FRAME_TYPE_DATAGRAM);
+                    builder.encode(&data);
+                }
+                debug_assert!(builder.len() <= builder.limit());
+                self.datagram = None;
```

Comment:
When we can't send a datagram, we might consider generating an event.  That event shouldn't contain the datagram, but it could include the length of the datagram and a reason (mostly just expiration).  That might be something to track in a new issue rather than doing it as part of this though.